### ISSUE TITLE
jax2tf shape polymorphism error for Haiku model

### ISSUE THREAD

**Issue by gnecula** [google/jax#6951]
The following code throws an `InconclusiveDimensionComputation` during the translation to TF of the gradient. 

```
    import os
    os.environ['HAIKU_FLATMAPPING'] = '0'
    import haiku as hk
    import tree

    def resnet(x, is_training):
      net = hk.nets.ResNet18(2)
      return net(x, is_training)

    resnet = hk.transform_with_state(resnet)
    inputs = np.random.random((5, 64, 64, 3))

    rng = jax.random.PRNGKey(42)

    params, states = resnet.init(rng, inputs, is_training=True)
    logits, states = resnet.apply(params, states, rng, inputs, is_training=True)

    resnet_apply = functools.partial(hk.without_apply_rng(resnet).apply, is_training=False)

    # Jax -> TF
    tf_fn = jax2tf.convert(resnet_apply,
                           polymorphic_shapes=[None, None, '(b, ...)'],
                           with_gradient=True)
    # Use TF gradient
    param_vars = tf.nest.map_structure(
        lambda param: tf.Variable(param, trainable=True),
        params)

    states_vars = tf.nest.map_structure(
        lambda state: tf.Variable(state, trainable=False), states)
    with tf.GradientTape() as tape:
      res = tf_fn(param_vars, states_vars, inputs)
      print("TF grad", tape.gradient(res, param_vars))
```

The traceback:
```
  File "/Users/necula/Source/jax/jax/experimental/jax2tf/jax2tf.py", line 286, in fun_vjp_jax
    return pullback_jax(out_cts_jax)
  File "/Users/necula/Source/jax/jax/_src/api.py", line 1881, in _vjp_pullback_wrapper
    ans = fun(*args)
  File "/Users/necula/Source/jax/jax/interpreters/ad.py", line 121, in unbound_vjp
    arg_cts = backward_pass(jaxpr, consts, dummy_args, cts)
  File "/Users/necula/Source/jax/jax/interpreters/ad.py", line 221, in backward_pass
    **eqn.params)
  File "/Users/necula/Source/jax/jax/_src/lax/lax.py", line 2700, in _add_transpose
    return [_unbroadcast(x_aval, t), _unbroadcast(y_aval, t)]
  File "/Users/necula/Source/jax/jax/_src/lax/lax.py", line 2278, in _unbroadcast
    if aval.shape == x_shape:
  File "/Users/necula/Source/jax/jax/experimental/jax2tf/shape_poly.py", line 212, in eq
    raise core.InconclusiveDimensionOperation(f"Dimension polynomial comparison '{self}' == '{other}' is inconclusive")
jax.core.InconclusiveDimensionOperation: Dimension polynomial comparison 'b' == '1' is inconclusive
```
